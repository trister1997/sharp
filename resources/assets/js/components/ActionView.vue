<template>
    <div class="SharpActionView">
        <template v-if="showErrorPage">
            <div class="container">
                <h1>Error {{errorPageData.status}}</h1>
                <p>{{errorPageData.message}}</p>
            </div>
        </template>
        <template v-else>
            <component v-if="barComp" :is="barComp" />
            <slot />
            <notifications position="top right" animation-name="slideRight" style="top:6rem" reverse>
                <template slot="body" slot-scope="{ item, close }">
                    <div class="SharpToastNotification" :class="`SharpToastNotification--${item.type||'info'}`" role="alert" data-test="notification">
                        <div class="SharpToastNotification__details">
                            <h3 class="SharpToastNotification__title mb-2">{{ item.title }}</h3>
                            <p v-if="!!item.text" class="SharpToastNotification__caption" v-html="item.text"></p>
                        </div>
                        <button class="SharpToastNotification__close-button" type="button" @click="close" data-test="close-notification">
                            <svg class="SharpToastNotification__icon" aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd">
                                <path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
                            </svg>
                        </button>
                    </div>
                </template>
            </notifications>
            <sharp-modal v-for="(modal,id) in mainModalsData" :key="id"
                v-bind="modal.props" @ok="modal.okCallback" @hidden="modal.hiddenCallback">
                {{modal.text}}
            </sharp-modal>
        </template>
    </div>
</template>

<script>
    import axios from 'axios';
    import { actionBarByContext } from './action-bar';
    import EventBus from './EventBus';
    import { api } from "../api";
    import SharpModal from './Modal';

    const noop=()=>{};

    export default {
        name:'SharpActionView',
        components: {
            SharpModal
        },

        provide() {
            return {
                actionsBus: new EventBus({name:'SharpActionsEventBus'}),
                axiosInstance: api
            }
        },

        props: {
            context: {
                type: String,
                required: true
            }
        },

        data() {
            return {
                mainModalsData: {},
                mainModalId: 0,
                showErrorPage: false,
                errorPageData: null
            }
        },
        computed: {
            barComp() {
                return actionBarByContext(this.context);
            },
        },
        methods: {
            showMainModal({ text, okCallback=noop, okCloseOnly, isError, ...sharedProps }) {
                const curId = this.mainModalId;
                const hiddenCallback = () => this.$delete(this.mainModalsData, curId);

                this.$set(this.mainModalsData,curId,{
                    props: {
                        ...sharedProps,
                        okOnly:okCloseOnly,
                        noCloseOnBackdrop:okCloseOnly,
                        noCloseOnEsc:okCloseOnly,
                        visible: true,
                        isError
                    },
                    okCallback, hiddenCallback,
                    text,
                });
                this.mainModalId++;
            }
        },
        created() {
            let { actionsBus, axiosInstance } = this._provided;

            actionsBus.$on('showMainModal', this.showMainModal);

            axiosInstance.interceptors.response.use(c=>c, error=>{
                let { response: {status, data}, config: { method } } = error;

                if(method==='get' && status === 404) {
                    this.showErrorPage = true;
                    this.errorPageData = {
                        status, message: data.message
                    }
                }
                return Promise.reject(error);
            });
        }
    }
</script>
