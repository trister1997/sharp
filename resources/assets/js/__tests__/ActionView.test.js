import Notifications from 'vue-notification';
import { shallowMount, createLocalVue, config } from '@vue/test-utils';
import ActionView from '../components/ActionView.vue';
import { createStub } from "./test-utils/vue-test-utils";

config.stubs['transition-group'] = false;

/* todo: fix tests on next release of vue-test-utils (1.0.0-beta.26)
    dynamic component aren't currently stubbed ( <component :is="{ ... }" /> )
*/

describe('action-view', ()=>{

    const ModalStub = {
        name:'TestModal',
        template:'<div><slot/></div>'
    };

    const localVue = createLocalVue();
    localVue.use(Notifications);

    function notify(wrapper, ...args) {
        wrapper.vm.$notify(...args);
        // remove notification global data-id to preserve unit
        wrapper.findAll('[data-id]').wrappers.forEach(wrapper=>wrapper.element.removeAttribute('data-id'));
    }

    function createWrapper(options={}) {
        return shallowMount(ActionView, {
            ...options,
            propsData: {
                context: 'form',
                ...options.propsData
            },
            computed: {
                barComp() {
                    return createStub(ActionView.computed.barComp.call(this));
                }
            },
            stubs: {
                'SharpModal': ModalStub
            },
            localVue
        })
    }

    test('can mount ActionView', ()=>{
        expect(createWrapper().html()).toMatchSnapshot();
    });

    test('can mount ActionView with notification', ()=>{
        const wrapper = createWrapper();
        notify(wrapper, { title:'TITLE', text:'MESSAGE', type:'TYPE' });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount ActionView with error page', ()=>{
        const wrapper = createWrapper({
            data: ()=>({
                showErrorPage: true,
                errorPageData: {
                    status: 404,
                    message: 'ERROR MESSAGE'
                }
            })
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount with multiple modals', ()=>{
        const wrapper = createWrapper();
        wrapper.vm.showMainModal({ text: 'Modal 1' });
        wrapper.vm.showMainModal({ text: 'Modal 2' });
        expect(wrapper.html()).toMatchSnapshot();
    });

    xtest('handle main modal events', ()=>{
        const wrapper = createWrapper();
        const modalOptions = { text: 'Modal 1', okCallback:jest.fn(), hiddenCallback:jest.fn() };
        wrapper.vm.showMainModal(modalOptions);

        let modal = wrapper.find(ModalStub);

        modal.vm.$emit('ok');
        expect(modalOptions.okCallback).toHaveBeenCalled();
    });
});
