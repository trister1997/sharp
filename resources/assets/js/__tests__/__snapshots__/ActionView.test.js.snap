// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action-view can mount ActionView 1`] = `
<div class="SharpActionView">
	<sharpactionbarform-stub></sharpactionbarform-stub>
	<notifications-stub group="" width="300" reverse="true" position="top right" classes="vue-notification" animationtype="css" animation="[object Object]" animationname="slideRight" speed="300" cooldown="0" duration="3000" delay="0"
	    max="Infinity" style="top: 6rem;"></notifications-stub>
</div>
`;

exports[`action-view can mount ActionView with error page 1`] = `
<div class="SharpActionView">
	<div class="container">
		<h1>Error 404</h1>
		<p>ERROR MESSAGE</p>
	</div>
</div>
`;

exports[`action-view can mount ActionView with notification 1`] = `
<div class="SharpActionView">
	<sharpactionbarform-stub></sharpactionbarform-stub>
	<notifications-stub group="" width="300" reverse="true" position="top right" classes="vue-notification" animationtype="css" animation="[object Object]" animationname="slideRight" speed="300" cooldown="0" duration="3000" delay="0"
	    max="Infinity" style="top: 6rem;"></notifications-stub>
</div>
`;

exports[`action-view can mount with multiple modals 1`] = `
<div class="SharpActionView">
	<sharpactionbarform-stub></sharpactionbarform-stub>
	<notifications-stub group="" width="300" reverse="true" position="top right" classes="vue-notification" animationtype="css" animation="[object Object]" animationname="slideRight" speed="300" cooldown="0" duration="3000" delay="0"
	    max="Infinity" style="top: 6rem;"></notifications-stub>
	<div visible="visible">
		Modal 1
	</div>
	<div visible="visible">
		Modal 2
	</div>
</div>
`;
